src/osp/Resource/load_tinygltf.cpp

```diff
@@ -0,0 +1,284 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "load_tinygltf.h"
+
+#include "ImporterData.h"
+#include "resources.h"
+
+#include "../logging.h"
+#include "../string_concat.h"
+
+#include <MagnumPlugins/TinyGltfImporter/TinyGltfImporter.h>
+#include <MagnumExternal/TinyGltf/tiny_gltf.h>
+
+#include <Magnum/Trade/ImageData.h>
+#include <Magnum/Trade/TextureData.h>
+#include <Magnum/Trade/MeshData.h>
+#include <Magnum/Trade/ObjectData3D.h>
+#include <Magnum/Trade/MeshObjectData3D.h>
+#include <Magnum/Trade/SceneData.h>
+
+#include <Corrade/PluginManager/Manager.h>
+
+using namespace osp;
+
+using Magnum::Trade::TinyGltfImporter;
+
+using Magnum::Trade::ImageData2D;
+using Magnum::Trade::TextureData;
+using Magnum::Trade::MeshData;
+using Magnum::Trade::MeshObjectData3D;
+using Magnum::Trade::MaterialData;
+using Magnum::Trade::ObjectData3D;
+using Magnum::Trade::SceneData;
+
+
+using Magnum::Trade::ObjectInstanceType3D;
+
+using Magnum::UnsignedInt;
+
+using PluginManager = Corrade::PluginManager::Manager<Magnum::Trade::AbstractImporter>;
+using Corrade::Containers::Optional;
+using Corrade::Containers::Pointer;
+
+using TinyGltfNodeExtras_t = std::vector<tinygltf::Value>;
+
+void osp::register_tinygltf_resources(Resources &rResources)
+{
+    rResources.data_register<TinyGltfNodeExtras_t>(restypes::gc_importer);
+}
+
+static void load_gltf(TinyGltfImporter &rImporter, ResId res, std::string_view name, Resources &rResources, PkgId pkg)
+{
+    ImporterData &rImportData = rResources.data_add<ImporterData>(restypes::gc_importer, res);
+
+    // Combine resource names. Maybe make this customizable
+    // ie: name = "dir/file.gltf" and resName = "mytexture"
+    // "dir/file.gltf:mytexture"
+    // "unnamed-[id]" is used as the resource name if it's empty
+    auto format_name = [tempStr = std::string{}, name]
+            (std::string_view resName, UnsignedInt id) mutable -> std::string const&
+    {
+        tempStr.clear();
+        if ( ! resName.empty())
+        {
+            string_append(tempStr, name, ":", resName);
+        }
+        else
+        {
+            // i don't like std::to_string but screw it, fix it later
```

Comment:
Same.

But I haven't gotten around to attempting to implement the in-place conversion function described in this paper: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1228r1.html